heat/core/dndarray.py

```diff
@@ -699,42 +694,51 @@ def __getitem__(self, key: Union[int, Tuple[int, ...], List[int, ...]]) -> DNDar
                 entries in the 0th dim refer to a single element. To handle this, the key is split
                 into the torch tensors for each dimension. This signals that advanced indexing is
                 to be used. """
-            key = manipulations.resplit(key.copy())
+            # NOTE: this gathers the entire key on every process!!
+            # TODO: remove this resplit!!
+            key = manipulations.resplit(key)
+            if key.larray.dtype in [torch.bool, torch.uint8]:
+                key = indexing.nonzero(key)
+
             if key.ndim > 1:
                 key = list(key.larray.split(1, dim=1))
                 # key is now a list of tensors with dimensions (key.ndim, 1)
                 # squeeze singleton dimension:
-                key = tuple(key[i].squeeze_(1) for i in range(len(key)))
+                key = list(key[i].squeeze_(1) for i in range(len(key)))
             else:
-                key = (key,)
+                key = [key]
             advanced_ind = True
         elif not isinstance(key, tuple):
             """ this loop handles all other cases. DNDarrays which make it to here refer to
                 advanced indexing slices, as do the torch tensors. Both DNDaarrys and torch.Tensors
                 are cast into lists here by PyTorch. lists mean advanced indexing will be used"""
             h = [slice(None, None, None)] * self.ndim
             if isinstance(key, DNDarray):
-                key = manipulations.resplit(key.copy())
-                h[0] = key.larray.tolist()
+                key = manipulations.resplit(key)
+                if key.larray.dtype in [torch.bool, torch.uint8]:
+                    h[0] = torch.nonzero(key.larray).flatten()  # .tolist()
+                else:
+                    h[0] = key.larray.tolist()
             elif isinstance(key, torch.Tensor):
-                h[0] = key.tolist()
+                if key.dtype in [torch.bool, torch.uint8]:
+                    # (coquelin77) i am not certain why this works without being a list. but it works...for now
+                    h[0] = torch.nonzero(key).flatten()  # .tolist()
```

Comment:
`flatten` --> `nonzero(as_tuple=True)`  or `squeeze_(1)`

Reply:
see above

Reply:
in place or oop doesnt make a large difference here, the oop will be cleaned up quickly anyway